src/packages/BooleanQueryTokenizer/BooleanQueryTokenizer.ts

```diff
@@ -0,0 +1,178 @@
+/**
+ * @module BooleanQueryTokenizer is responsible for tokenizing boolean query language. Also
+ * it provides a method `extractWordsAndPhrases` that returns bag of phrases and words.
+ *
+ * Glossary:
+ * A _token_ is a pair consisting of a token name and matched _lexeme_.
+ * The token name represents a kind of lexical unit, e.g.,
+ * particular keyword, phrase, operator, separator, etc.
+ *
+ * A _lexeme_ is a sequence of characters in the input string that matches the pattern for
+ * a token and is identified by the lexical analyzer as an instance of that token.
+ *
+ * A _pattern_ is a description of the form that the lexemes of a token may take.
+ * In the case of a keyword as a token, the pattern is just the sequence of characters that
+ * form the keyword. For identifiers and some other tokens, the pattern is more complex
+ * structure that is matched by many strings.
+ */
```

Comment:
Great to have all this documentation!
Small question, what is `lexeme` stands for?

Reply:
It's a commonly used term in CS/parsing field. I believe it came from linguistics https://en.wikipedia.org/wiki/Lexeme

Reply:
we learn something new everyday. Or is it relearning it 🤔

Reply:
> we learn something new everyday.
true :D